File: number-of-provinces/number-of-provinces.cpp
#include<bits/stdc++.h>
using namespace std;

class DisjointSet
{
    int *rank, *parent, n;
 
public:
    DisjointSet(int n)
    {
        rank = new int[n];
        parent = new int[n];
        this->n = n;
        makeSet();
    }
 
    void makeSet()
    {
        for (int i = 0; i < n; i++)
        {
            parent[i] = i;
        }
    }
    
    // Returns the parent/ set of the given item x
    int find(int x)
    {
        if (parent[x] != x)
        {
 
            parent[x] = find(parent[x]);
        }
 
        return parent[x];
    }
    
    //Merges two sets
    void Union(int x, int y)
    {
 
        int xset = find(x);
        int yset = find(y);
 
        if (xset == yset) 
            return;
 
        if (rank[xset] < rank[yset])
        {
            parent[xset] = yset;
        }
        else if (rank[xset] > rank[yset])
        {
            parent[yset] = xset;
        }
 
        else
        {
            parent[yset] = xset;
            rank[xset] = rank[xset] + 1;
        }
    }
};

class Solution {
public:
    int findCircleNum(vector<vector<int>>& arr) 
    {
        int n=arr.size();
        DisjointSet obj(n);
        map<int,int>mp;
        int ct=0;
        for(int i=0; i<n; i++)
        {
            for(int j=0; j<n; j++)
            {
                if(i!=j&&arr[i][j]==1&&obj.find(i)!=obj.find(j))
                    obj.Union(i,j);
            }
        }
        for(int i=0; i<n; i++)
        {
            if(!mp[obj.find(i)])
            {
                ct++;
                mp[obj.find(i)]=1;
            }
        }
        return ct;
        
    }
};

File: number-of-provinces/number-of-provinces.py
class DisjointSet:
    def __init__(self,n):
        self.rank=[0]*n
        self.parent=[0]*n
        self.n=n
        self.makeSet()

    def makeSet(self):
        for i in range(self.n):
            self.parent[i]=i

    def find(self,x) -> int:
        if(self.parent[x]==x):
            return x
        self.parent[x]=self.find(self.parent[x])
        return self.parent[x]

    def Union(self,x,y):
        xset=self.find(x)
        yset=self.find(y)
        if(xset==yset):
            return
        if(self.rank[xset]<self.rank[yset]):
            self.parent[xset]=yset

        elif(self.rank[xset]>self.rank[yset]):
            self.parent[yset]=xset
        else:
            self.parent[yset]=xset
            self.rank[xset]+=1

class Solution:
    def findCircleNum(self, arr: List[List[int]]) -> int:
        n , ct =len(arr) , 0
        obj=DisjointSet(n)
        mp={}
        for i in range(n):
            for j in range(n):
                if(i!=j and arr[i][j]==1 and obj.find(i)!=obj.find(j)):
                    obj.Union(i,j)
    
        for i in range(n):
            if(obj.find(i) not in mp):
                ct+=1
                mp[obj.find(i)]=1

        return ct   
        